### ISSUE TITLE
Add Annotation#types, #vars, #methods

### ISSUE THREAD

**Issue by vladfaust** [crystal-lang/crystal#7274]
These methods would allow to iterate over all types/vars/methods which have the annotation, e.g:

```crystal
annotation Foo
end

@[Foo]
class Bar
end

{% pp Foo.types %} # => [Bar]
```

/cc @Blacksmoke16, this could be useful in [Athena](https://github.com/Blacksmoke16/athena), right?

I also can find it useful in ORM implementations.

**Comment by Blacksmoke16** [comment]
Yes, this would allow me to just get a list of all methods that are tagged as a route action.  Currently I am requiring the controller classes to inherit from a predefined class, then iterating over the subclasses of that parent class, then filtering the methods down to just ones with the annotation.

This would be a bit cleaner, and allow the user to define their own inheritance structure.

**Comment by vladfaust** [comment]
@Blacksmoke16 there is an alternative possible solution: https://forum.crystal-lang.org/t/type-all-includers/254

**Comment by Blacksmoke16** [comment]
@vladfaust What would your use case be for wanting an array of _ALL_ methods/instance vars?

Would something like

```crystal
annotation Foo
end

class Klass
  @[Foo]
  def one
    1
  end
  
  @[Foo]
  def two
    2
  end

  def three
    3
  end
end

pp {{Klass.methods(Foo)}} # => [one, two]
```
Where `methods` on `TypeNode` optionally takes an `AnnotationType`, and only returns methods with that annotation?  This is already currently possible, would be fore sure be cleaner.

**Comment by HertzDevil** [comment]
If https://github.com/crystal-lang/crystal/pull/7648#issuecomment-598167731 works for types, do we still need this feature for instance / class variables and defs, or are there similar workarounds for them?

**Comment by Blacksmoke16** [comment]
@HertzDevil You can basically use the same workaround but with `@type.methods`.  Only thing this feature would allow for is like getting an `Array(Def)` of _all_ methods that have a given annotation irrespective of a specific type.  However I'm not sure of how that be useful/why you would need that.

**Comment by kalinon** [comment]
I also realized today that the workaround im used to using (`Object.all_subclasses`) misses any modules with annotations as well.
